clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -908,311 +628,781 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
   }
 
-  return AllArgsAreValid;
-}
+public:
+  SyclKernelBodyCreator(Sema &S, SyclKernelDeclCreator &DC,
+                        CXXRecordDecl *KernelObj,
+                        FunctionDecl *KernelCallerFunc)
+      : SyclKernelFieldHandler(S), DeclCreator(DC),
+        KernelObjClone(createKernelObjClone(S.getASTContext(),
+                                            DC.getKernelDecl(), KernelObj)),
+        VarEntity(InitializedEntity::InitializeVariable(KernelObjClone)),
+        KernelObj(KernelObj), KernelCallerFunc(KernelCallerFunc) {
+    markParallelWorkItemCalls();
+
+    Stmt *DS = new (S.Context) DeclStmt(DeclGroupRef(KernelObjClone),
+                                        SourceLocation(), SourceLocation());
+    BodyStmts.push_back(DS);
+    DeclRefExpr *KernelObjCloneRef = DeclRefExpr::Create(
+        S.Context, NestedNameSpecifierLoc(), SourceLocation(), KernelObjClone,
+        false, DeclarationNameInfo(), QualType(KernelObj->getTypeForDecl(), 0),
+        VK_LValue);
+    MemberExprBases.push_back(KernelObjCloneRef);
+  }
+
+  ~SyclKernelBodyCreator() {
+    CompoundStmt *KernelBody = createKernelBody();
+    DeclCreator.setBody(KernelBody);
+  }
+
+  void handleSyclAccessorType(FieldDecl *FD, QualType Ty) final {
+    handleSpecialType(FD, Ty);
+  }
+
+  void handleSyclAccessorType(const CXXBaseSpecifier &BS, QualType Ty) final {
+    // TODO: Creates init sequence and inits special sycl obj
+  }
+
+  void handleSyclSamplerType(FieldDecl *FD, QualType Ty) final {
+    handleSpecialType(FD, Ty);
+  }
+
+  void handleSyclStreamType(FieldDecl *FD, QualType Ty) final {
+    const auto *StreamDecl = Ty->getAsCXXRecordDecl();
+    // Hmm okay, we created "stream" kernel argument and did copy initialization
+    // but why?
+    // We also don't have SEMA tests for stream, only pretty small CodeGen test
+    createExprForStructOrScalar(FD);
+    createSpecialMethodCall(StreamDecl, MemberExprBases.back(), InitMethodName,
+                            FD);
+    createSpecialMethodCall(StreamDecl, MemberExprBases.back(),
+                            FinalizeMethodName, FD);
+  }
+
+  void handleSyclStreamType(const CXXBaseSpecifier &BS, QualType Ty) final {
+    // TODO: Creates init/finalize sequence and inits special sycl obj
+  }
+
+  void handlePointerType(FieldDecl *FD, QualType ArgTy) final {
+    createExprForStructOrScalar(FD);
+  }
+
+  void handleStructType(FieldDecl *FD, QualType Ty) final {
+    createExprForStructOrScalar(FD);
+  }
+
+  void handleScalarType(FieldDecl *FD, QualType Ty) final {
+    createExprForStructOrScalar(FD);
+  }
+
+  void enterStruct(const CXXRecordDecl *, FieldDecl *FD) final {
+    MemberExprBases.push_back(BuildMemberExpr(MemberExprBases.back(), FD));
+  }
+
+  void leaveStruct(const CXXRecordDecl *, const FieldDecl *FD) final {
+    MemberExprBases.pop_back();
+  }
+
+  void enterStruct(const CXXRecordDecl *RD, const CXXBaseSpecifier &BS) final {
+    // TODO : Implement
```

Comment:
Probably remove these two since they are empty, but if necessary, document the 'how' in the bug report.

Reply:
I can't remove them, it will be compilation error.

> but if necessary, document the 'how' in the bug report.
What do you mean?

Reply:
It shouldn't be... the visitor base should handle the 'do nothing case'.  What compilation error are you seeing?

>> What do you mean?
I just mean don't leave a "TODO: Implement", either implement it or capture what needs to happen/how you'll solve it in a bug report.

Reply:
If I define only one version of `enterStruct` (or `leaveStruct` whatever) in derived class it seems that the compiler cannot find other one:
```
llvm/clang/lib/Sema/SemaSYCL.cpp:713:72: error: no matching function for call to ‘{anonymous}::SyclKernelBodyCreator::enterStruct(clang::CXXRecordDecl*&, const clang::CXXBaseSpecifier&)
’
   (void)std::initializer_list<int>{(handlers.enterStruct(Owner, Parent), 0)...};
```
no matter that the base class defines it.

Reply:
Ah, right, name hiding stuff again. can you just do these as a 'using' statement then? "using <basename>::enterStruct".

Reply:
Okay, done.